common/frame_session.go

```diff
@@ -604,21 +669,69 @@ func (fs *FrameSession) onExecutionContextDestroyed(execCtxID cdpruntime.Executi
 	delete(fs.contextIDToContext, execCtxID)
 }
 
+func (fs *FrameSession) addClearWait(c chan bool, frameID string) {
+	fs.logger.Debugf("FrameSession:addClearWait", "navigateFrame adding wait chan fid:%s", frameID)
+
+	cPrv, ok := fs.getClearWaitChan(frameID)
+
+	if ok {
+		select {
+		case <-cPrv:
+			// onExecutionContextsCleared should be the only thing
+			// closing the channel and is also in charge of deleting
+			// it from the map, so what has closed it without deleting it?
+			fs.logger.Errorf("FrameSession:addClearWait", "channel already closed fid:%s", frameID)
+		case <-time.After(time.Second * 5):
+			// Unknown: Why have we got an open channel which has not been closed?
+			// 			If it's a valid reason then change this to a Debugf.
+			fs.logger.Errorf("FrameSession:addClearWait", "force closing chan fid:%s", frameID)
+			close(cPrv)
+		}
+	}
+
+	fs.clearWaitChansMu.Lock()
+	defer fs.clearWaitChansMu.Unlock()
+
+	fs.clearWaitChans[frameID] = c
+}
+
+func (fs *FrameSession) getClearWaitChan(frameID string) (chan bool, bool) {
+	fs.clearWaitChansMu.Lock()
+	defer fs.clearWaitChansMu.Unlock()
+
+	c, ok := fs.clearWaitChans[frameID]
+
+	return c, ok
+}
+
 func (fs *FrameSession) onExecutionContextsCleared() {
 	fs.logger.Debugf("FrameSession:onExecutionContextsCleared",
 		"sid:%v tid:%v", fs.session.ID(), fs.targetID)
 
 	fs.contextIDToContextMu.Lock()
-	defer fs.contextIDToContextMu.Unlock()
 
+	var fid string
 	for _, context := range fs.contextIDToContext {
+		fid = context.Frame().ID()
```

Comment:
Iterating over a map is unstable, and we'll only get the last iteration's frame ID here.

Couldn't that leave other channels opened, and lead to timeouts in `addClearWait()`?